python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -224,74 +224,82 @@ def resolve(self, spec):
         uri = spec.getUri()
         options = spec.getOptions()
         logger.info(f"Resolving model {name} from {uri}")
-        model_uri, tags, params = _get_model_info(uri, self.tracking_client)
+        parsed = urlparse(uri)
+        if not parsed.scheme:
+            raise ValueError("Scheme must be mlflow. How did you get here?")
+        parts = parsed.path.strip("/").split("/", 1)
+        if self.is_model(parts[0]):
+            model_uri, run = self.get_model_version(*parts)
+        else:
+            model_uri, run = self.get_run(*parts)
         spec = MlflowModelSpec(
-            model_uri, tags, params, self.mlflow_tracking_uri, options=options
+            model_uri,
+            run.data.tags,
+            run.data.params,
+            self.mlflow_tracking_uri,
+            options=options,
         )
         func_name = codegen_from_spec(self._spark, spec, name)
         model = self._jvm.ai.eto.rikai.sql.model.SparkUDFModel(
             name, uri, func_name
         )
         return model
 
-
-def _get_model_info(uri: str, client: MlflowClient) -> str:
-    """Transform the rikai model uri to something that mlflow understands"""
-    parsed = urlparse(uri)
-    try:
-        client.get_registered_model(parsed.hostname)
-        return _parse_model_ref(parsed, client)
-    except MlflowException:
-        return _parse_runid_ref(parsed, client)
-
-
-def _parse_model_ref(parsed: ParseResult, client: MlflowClient):
-    model = parsed.hostname
-    path = parsed.path.lstrip("/")
-    if path.isdigit():
-        mv = client.get_model_version(model, int(path))
-        run = client.get_run(mv.run_id)
-        return (
-            "models:/{}/{}".format(model, path),
-            run.data.tags,
-            run.data.params,
-        )
-    if not path:
-        stage = "none"  # TODO allow setting default stage from config
-    else:
-        stage = path.lower()
-    results = client.get_latest_versions(model, stages=[stage])
-    if not results:
-        raise SpecError(
-            "No versions found for model {} in stage {}".format(model, stage)
+    def is_model(self, entity_reference):
+        """
+        Returns True is the given mlflow entity reference is a registered model
+        """
+        try:
+            self.tracking_client.get_registered_model(entity_reference)
+            return True
+        except Exception:
```

Comment:
a `SocketTimeout` will return False as well. Is it desired?

Reply:
THis is  trying to figure out whether the input is a model name or a run id. In the case of a sockettimeout it would think it's a runid and try to do get_run(entity_reference) which would also fail. 

Three alternatives:

have mlflow+models:/<model_name>/<version>
or mlflow:/models/<model_name>/<version>
or just not support direct references to a runid (only support mlflow model registry references)

Reply:
can we register `MLFlowRegistery` with both `rikai.ml.sql.registry.models.impl` AND  `rikai.ml.sql.registry.runs.impl` ?

Reply:
i guess my main desire here is that users can use `rikai` in a way that they dont need to convert different URI pattern in mind, for the sake of better UX / `Principle of least astonishment` ;)

Reply:
Yeah, I thought about that as well. Very reasonable points.

I decided against surfacing the mlflow:/ and runs:/ schemes to Rikai users because:
1. models:/ and runs:/ make sense in the context of mlflow but are ambiguous in wider contexts.
2. I see the mlflow style "models:/model" and "runs:/run_id" as implementation details internal to mlflow. 
3. I want our users thinking in terms of model names, versions, and at most run_id (only when absolutely necessary) 
4. Having both models:/ and runs:/ means you can have them both as the default uri root.
5. It's likely that SQL ML users won't really be all that familiar with the implementation details of Mlflow so they probably don't even know the "models:/" and "runs:/" URI formats.

Reply:
makes sense. Guess we can see user feedbacks on this later.  LGTM overall.